modules/schema/src/main/java/org/apache/ignite/internal/schema/SchemaManager.java

```diff
@@ -300,27 +274,19 @@ private CompletableFuture<SchemaDescriptor> tableSchema(UUID tblId, String table
 
         CompletableFuture<SchemaDescriptor> fut = new CompletableFuture<>();
 
-        if (checkSchemaVersion(tblId, schemaVer)) {
-            SchemaDescriptor desc = searchSchemaByVersion(tblId, schemaVer);
+        SchemaRegistry registry = registriesVv.latest().get(tblId);
 
-            if (desc == null) {
-                return getSchemaDescriptor(schemaVer, tblCfg);
-            } else {
-                fut.complete(desc);
-                return fut;
-            }
+        // TODO: We can remove it after: restore all schemes on recovery https://issues.apache.org/jira/browse/IGNITE-18066
+        // Probably all near stuff must be improved too.
+        if (registry.lastSchemaVersion() > schemaVer) {
+            return getSchemaDescriptor(schemaVer, tblCfg);
         }
 
         IgniteTriConsumer<Long, Map<UUID, SchemaRegistryImpl>, Throwable> schemaListener = (token, regs, e) -> {
             if (schemaVer <= regs.get(tblId).lastSchemaVersion()) {
-                getSchemaDescriptor(schemaVer, tblCfg)
-                        .whenComplete((desc, th) -> {
-                            if (th != null) {
-                                fut.completeExceptionally(th);
-                            } else {
-                                fut.complete(desc);
-                            }
-                        });
+                SchemaRegistry registry0 = registriesVv.latest().get(tblId);
+
+                fut.complete(registry0.schemaCached(schemaVer));
```

Comment:
could you pls add an assertion that `registry0.schemaCached(schemaVer)` didn't return `null`?